share/h2o/mruby/acl.rb

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) 2016 DeNA Co., Ltd., Ichito Nagata
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+module H2O
+
+  module ACL
+
+    def acl(&block)
+      $ACL += 1
+      return ACLBlock.new._to_handler(&block)
+    end
+
+    class ACLHandler
+      def initialize(p)
+        @proc = p
+      end
+      def call(env)
+        return @proc.call(env)
+      end
+    end
+
+    class ACLBlock
+
+      class ConditionalHandler
+        def initialize(handler, cond)
+          @handler = handler
+          @cond = cond
+        end
+
+        def satisfy?(env)
+           return @cond.nil? || MatchingBlock.new(env).instance_eval(&@cond)
+        end
+
+        def call(env)
+          return @handler.call(env) if satisfy?(env)
+          return [399, {}, []]
+        end
+      end
+
+      def initialize
+        @acl = []
+      end
+
+      def _to_handler(&block)
+        returned = instance_eval(&block)
+        return ACLHandler.new(lambda {|env|
+          @acl.each {|ac|
+            return ac.call(env) if ac.satisfy?(env)
+          }
+          return [399, {}, []] if returned.nil?
+          return returned.call(env)
+        })
```

Comment:
Would it be possible to merge `ACLHandler` and `ACLBlock` into one class?

The reason I ask is because `ACLHandler#call` invoking the proc returned by `ACLHandler#_to_handler` for every request looks like a waste of CPU cycles (and because I cannot see why we need such indirection).